sycl/source/detail/usm/usm_impl.cpp

```diff
@@ -226,5 +226,70 @@ void *aligned_alloc(size_t Alignment, size_t Size, const queue &Q, alloc Kind) {
   return aligned_alloc(Alignment, Size, Q.get_device(), Q.get_context(), Kind);
 }
 
+// Pointer queries
+/// Query the allocation type from a USM pointer
+/// Returns alloc::host for all pointers in a host context.
+///
+/// @param ptr is the USM pointer to query
+/// @param ctxt is the sycl context the ptr was allocated in
+alloc get_pointer_type(const void *Ptr, const context &Ctxt) {
+  // Everything on a host device is just system malloc so call it host
+  if (Ctxt.is_host())
+    return alloc::host;
+
+  std::shared_ptr<detail::context_impl> CtxImpl = detail::getSyclObjImpl(Ctxt);
+  pi_context C = CtxImpl->getHandleRef();
+  pi_usm_type AllocTy;
+
+  // query type using PI function
+  PI_CALL(piextUSMGetMemAllocInfo)(C, Ptr, PI_MEM_ALLOC_TYPE,
+                                   sizeof(pi_usm_type), &AllocTy, nullptr);
+
+  alloc ResultAlloc = alloc::unknown;
+  if (AllocTy == PI_MEM_TYPE_HOST) {
+    ResultAlloc = alloc::host;
+  }
+  else if (AllocTy == PI_MEM_TYPE_DEVICE) {
+    ResultAlloc = alloc::device;
+  }
+  else if (AllocTy == PI_MEM_TYPE_SHARED) {
+    ResultAlloc = alloc::shared;
+  }
+
+  return ResultAlloc;
+}
+
+/// Queries the device against which the pointer was allocated
+///
+/// @param ptr is the USM pointer to query
+/// @param ctxt is the sycl context the ptr was allocated in
+device get_pointer_device(const void *Ptr, const context &Ctxt) {
+  // Just return the host device in the host context
+  if (Ctxt.is_host())
+    return Ctxt.get_devices()[0];
+
+  // Check if ptr is a host allocation
+  if (get_pointer_type(Ptr, Ctxt) == alloc::host)
+    return device();
+
+  std::shared_ptr<detail::context_impl> CtxImpl = detail::getSyclObjImpl(Ctxt);
+  pi_context C = CtxImpl->getHandleRef();
+  pi_device DeviceId;
+
+  // query device using PI function
+  PI_CALL(piextUSMGetMemAllocInfo)(C, Ptr, PI_MEM_ALLOC_DEVICE,
+                                   sizeof(pi_device), &DeviceId, nullptr);
+
+  auto Devs = Ctxt.get_devices();
+
+  for (auto D : Devs) {
```

Comment:
for (const std::shared_ptr<detail::device_impl> &Dev : Ctxt.get_devices())

Reply:
get_devices() returns vector_class<device> not device_impl?

Reply:
My snippet is incorrect. get_devices returns vector_class<sycl::device>. New snippet:

for (const device &Dev : Ctxt.get_devices()) {
  if (detail::getSyclObjImpl(Dev)->getHandleRef() == DeviceId)
     return D;
}

The problem with calling device.get() is that it increases reference counter of cl_device_id object.

Reply:
Makes sense.